Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,89 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a hook that allows managed code
+to run before the main application's entry point.
+
+## Motivation
+
+This would allow hosting providers to put configuration and policy in
+a single location, including settings that potentially influence load
+behavior of the main entry point such as the AssemblyLoadContext
+behavior. This could be used to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook needs to be separate from the
+entry point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+Environment variables or `<appname>.runtimeconfig.json` can be used to
+specify a managed assembly and type that contains an `Initialize`
+method.
+
+```
+DOTNET_MANAGED_HOST_ASSEMBLY=/path/to/ManagedHost.dll
+DOTNET_MANAGED_HOST_TYPE=ManagedHostNamespace.ManagedHostType
+```
+
+```
+{
+    "runtimeOptions": {
+        "managedHostAssembly": "/path/to/ManagedHost.dll",
+        "managedHostType": "ManagedHostNamespace.ManagedHostType"
+    }
+}
+```
+
+The environment variables, if set, take precedence over the
+`<appname>.runtimeconfig.json` settings. These settings result in
+`ManagedHostType.Initialize()` being called in hostpolicy, before the
+main assembly is loaded.
+
+Specifically, hostpolicy starts up coreclr and sets up a new AppDomain
+with `ManagedHost.dll` on the TPA list. If it read these settings from
+environment variables, it clears them so that child processes do not
```

Comment:
I think that clearing the variable may or may not be the desired behavior.  It depends on what folks are using this for.

Reply:
`DOTNET_ROOT` environment variable is similar to this - it has similar problems. We are not clearing that one either.

Reply:
+1 I think we should leave them as is.  The developer can either use runtimeconfig or manually not inherit the env block on new process launches if desired.

Reply:
I changed the default to "do not clear" and pointed out the ramifications: https://github.com/dotnet/core-setup/pull/4421/commits/f5a3acfb07592c236062646ba4a6fee9d07164b6.
I think it makes sense to let the managed code decide what to do here.